prototyping/Luau/OpSem.agda

```diff
@@ -1,18 +1,62 @@
-open import Agda.Builtin.Float using (Float; primFloatPlus; primFloatMinus; primFloatTimes; primFloatDiv)
+open import Agda.Builtin.Float using (Float; primFloatPlus; primFloatMinus; primFloatTimes; primFloatDiv; primFloatEquality; primFloatLess; primFloatInequality)
+open import Agda.Builtin.Bool using (Bool; true; false)
+open import Utility.Bool using (not; _or_; _and_)
+open import Agda.Builtin.Nat using (_==_)
 open import FFI.Data.Maybe using (just)
 open import Luau.Heap using (Heap; _≡_⊕_↦_; _[_]; function_is_end)
 open import Luau.Substitution using (_[_/_]ᴮ)
-open import Luau.Syntax using (Expr; Stat; Block; nil; addr; var; function_is_end; _$_; block_is_end; local_←_; _∙_; done; return; name; fun; arg; binexp; BinaryOperator; +; -; *; /; number)
-open import Luau.Value using (addr; val; number)
-
-evalBinOp : Float → BinaryOperator → Float → Float
-evalBinOp x + y = primFloatPlus x y
-evalBinOp x - y = primFloatMinus x y
-evalBinOp x * y = primFloatTimes x y
-evalBinOp x / y = primFloatDiv x y
+open import Luau.Syntax using (Expr; Stat; Block; nil; addr; var; function_is_end; _$_; block_is_end; local_←_; _∙_; done; return; name; fun; arg; binexp; BinaryOperator; +; -; *; /; <; >; ≡; ≅; ≤; ≥; ∧; ∨; number)
+open import Luau.Value using (addr; val; number; Value; bool)
+open import Luau.RuntimeType using (RuntimeType; valueType)
+
+evalNumOp : Float → BinaryOperator → Float → Value
+evalNumOp x + y = number (primFloatPlus x y)
+evalNumOp x - y = number (primFloatMinus x y)
+evalNumOp x * y = number (primFloatTimes x y)
+evalNumOp x / y = number (primFloatDiv x y)
+evalNumOp x < y = bool (primFloatLess x y)
+evalNumOp x > y = bool (primFloatLess y x)
+evalNumOp x ≡ y = bool (primFloatEquality x y)
+evalNumOp x ≅ y = bool (primFloatInequality x y)
+evalNumOp x ≤ y = bool ((primFloatLess x y) or (primFloatEquality x y))
+evalNumOp x ≥ y = bool ((primFloatLess y x) or (primFloatEquality x y))
+evalNumOp x ∧ y = number x
+evalNumOp x ∨ y = number x
+
+evalEqOp : Value → Value → Value
+evalEqOp Value.nil Value.nil = bool true
+evalEqOp (addr x) (addr y) = bool (x == y)
+evalEqOp (number x) (number y) = bool (primFloatEquality x y)
+evalEqOp (bool true) (bool y) = bool y
+evalEqOp (bool false) (bool y) = bool (not y)
+evalEqOp _ _ = bool false
+
+evalNeqOp : Value → Value → Value
+evalNeqOp Value.nil Value.nil = bool false
+evalNeqOp (addr x) (addr y) = bool (not (x == y))
+evalNeqOp (number x) (number y) = bool (primFloatInequality x y)
+evalNeqOp (bool true) (bool y) = bool (not y)
+evalNeqOp (bool false) (bool y) = bool y
+evalNeqOp _ _ = bool true
+
+coerceToBool : Value → Bool
+coerceToBool Value.nil = false
+coerceToBool (addr x) = true
+coerceToBool (number x) = true
+coerceToBool (bool x) = x
+
+eval∨ : Value → Value → Value
+eval∨ x y with coerceToBool x
+eval∨ x y | false = y
+eval∨ x y | true = x
+
+eval∧ : Value → Value → Value
+eval∧ x y with coerceToBool x
+eval∧ x y | true = y
+eval∧ x y | false = x
```

Comment:
I wonder if we could/should roll these all together and just have one evalBinOp, to make the opsem simpler?